src/table.py

```diff
@@ -1462,22 +1487,33 @@ def _get_header(self, y_tolerance=3):
         page = self.page
         y_delta = y_tolerance
 
-        def top_row_is_bold(bbox):
-            """Check if row 0 has bold text anywhere.
+        def top_row_bg_color(self):
+            """
+            Compare top row background color with color of same-sized bbox
+            above. If different, return True indicating that the original
+            table top row is already the header.
+            """
+            bbox0 = Rect(self.rows[0].bbox)
+            bboxt = bbox0 + (0, -bbox0.height, 0, -bbox0.height)  # area above
+            top_color0 = page.get_pixmap(clip=bbox0).color_topusage()[1]
+            top_colort = page.get_pixmap(clip=bboxt).color_topusage()[1]
+            if top_color0 != top_colort:
+                return True  # top row is header
+            return False
 
-            If this is true, then any non-bold text in lines above disqualify
-            these lines as header.
+        def row_has_bold(bbox):
+            """Check if a row contains some bold text.
 
-            bbox is the (potentially repaired) row 0 bbox.
+            If e.g. true for the top row, then it will be used as (internal)
+            column header row if any of the following is true:
+            * the previous (above) text line has no bold span
+            * the second table row text has no bold span
 
-            Returns True or False
+            Returns True if any spans are bold else False.
             """
-            for b in page.get_text("dict", flags=TEXTFLAGS_TEXT, clip=bbox)["blocks"]:
-                for l in b["lines"]:
-                    for s in l["spans"]:
-                        if s["flags"] & 16:
-                            return True
-            return False
+            blocks = page.get_text("dict", flags=TEXTFLAGS_TEXT, clip=bbox)["blocks"]
+            spans = [s for b in blocks for l in b["lines"] for s in l["spans"]]
+            return any([bool(s["flags"] & 16) for s in spans])  # pylint: disable=R1729
```

Comment:
Can we use `TEXT_FONT_BOLD` here instead of `16`?

Reply:
Actually, the suggestion R1729 does not make much sense, because making a generator is unnecessary overhead for these small numbers of checks.
But I will replace this with a normal loop from which an early return exit is possible. More code of course, but what don't we do for the satisfaction of pylint?
The symbolic names for the hardcoded numbers are certainly reasonable.

Reply:
I can't see the problem with using generator expressions.

* According to https://peps.python.org/pep-0289/, compared to list comprehensions, performance with generator expressions is "roughly comparable for small to mid-sized data sets". And they are faster as the data set gets larger.

* The code using generator expressions is slightly smaller than the code that uses list comprehensions, and avoids the pylint warning.

* The code using generator expressions is much simpler than using an explicit loop.

* I don't know whether the explicit loop might be faster or not. I suspect any difference (whether faster or slower) would not be significant for PyMuPDF.

However if you really want to avoid generator expressions, that's ok by me too!